### PR TITLE
AML-21 drift detector is too sensitive

### PR DESCRIPTION
**Pull request by tstreule**
data drift is a concept where data in production differs too much from the training data (it drifted away). That’s obviously a problem for performance. We have a data drift detector that should warn for this, but it’s overly sensitive. It warns for drift even on training data.

### CODE REVIEW

**Comment by tstreule** [comment]
Please use the following commit message for merging:

```
AML-21 drift detector is too sensitive (#17)

* Add test case for testing drift warnings

* Use custom DataDriftWarnings

* Use`warnings` module instead of `logging` (in `Amplo/AutoML/DriftDetector.py`)

* Add `DriftDetector._check_distributions` to `DriftDetector.check` (in `Amplo/AutoML/DriftDetector.py`)

* Test sensibility of DriftDetector by asserting that no `DataDriftWarning` occurred

* Bugfix `Amplo/Utils/utils.py::hist_search`
```

File: Amplo/AutoML/DriftDetector.py [hunk 1]
```diff
@@ -98,13 +103,10 @@ def check_output(self, model, data: pd.DataFrame, add: bool = False):
...
-        while no_drift:
-            for value in prediction:
-                ind = hist_search(x, value)
-                if ind == -1 or y[ind] <= 0:
-                    logging.warning(f"[AutoML] Output drift detected!")
-                    no_drift = False
+        for value in prediction:
+            ind = hist_search(x, value)
+            if ind == -1 or y[ind] <= 0:
+                warnings.warn(DataDriftWarning("[AutoML] Output drift detected!"))
```
Comment (nielsuit227) [inline Amplo/AutoML/DriftDetector.py:109]: This has the potential to give a warning for every output, which is annoying when using predict on multiple samples, which is the most common usecase.

File: Amplo/Utils/utils.py [hunk 2]
```diff
@@ -13,14 +17,46 @@ def get_model(model_str, **args):
...
+        logging.debug(f'No bin (index) found for value {value}. Array(Min: {array[0]}, Max: {array[-1]})')
+        return -1
+
+    # Initialize min and max bin index
     low = 0
     high = len(array) - 1
-    while low < high:
+
+    # Bin search
+    countdown = 1000
```
Comment (nielsuit227) [inline Amplo/Utils/utils.py:45]: 2 ^ 1000 = an array of length 10e+300. We'll be dealing with max 1 trillion samples, which would be 2 ^ 30, so you can make this 30.

File: tests/Unit/AutoML/test_drift_detector.py [hunk 3]
```diff
@@ -1,46 +1,128 @@
...
+
+    return df, dict(cat_cols=cat_df.columns, num_cols=num_df.columns)
+
+
+def make_cat_data(num_samples: int, cat_choices=None):
+    return make_data(num_samples, cat_choices=cat_choices, num_dists=False)
+
+
+def make_num_data(num_samples: int, num_dists=None):
+    return make_data(num_samples, cat_choices=False, num_dists=num_dists)
```
Comment (nielsuit227) [inline tests/Unit/AutoML/test_drift_detector.py:97]: Can you put this in some test util? This is also useful for DataProcessor and FeatureProcessor :)

**Review by nielsuit227 — CHANGES_REQUESTED** [review]

File: Amplo/Utils/utils.py (hunk 2 above)
Comment (tstreule) [reply Amplo/Utils/utils.py]: Sure thing! 
commit 0537dbc66ff23b3b612f3ed5d2a637b2d19b281d

File: tests/Unit/AutoML/test_drift_detector.py (hunk 3 above)
Comment (tstreule) [reply tests/Unit/AutoML/test_drift_detector.py]: Sure thing!
commit 0537dbc66ff23b3b612f3ed5d2a637b2d19b281d

File: Amplo/AutoML/DriftDetector.py (hunk 1 above)
Comment (tstreule) [reply Amplo/AutoML/DriftDetector.py]: Good point!
commit 71deb186926423e26675db285d4db3ed1f88e870

**Review by nielsuit227 — APPROVED** [review]

**Pull request merged**